src/deriveNativeScriptHash.c

```diff
@@ -0,0 +1,464 @@
+
+// Helper functions
+
+#define TRACE_WITH_CTX(message, ...) TRACE(message "level = %u, remaining scripts = %u", ##__VA_ARGS__, ctx->level, ctx->complexScripts[ctx->level].remainingScripts)
+#define SAFE_SPRINTF(str, format, ...) {\
+		explicit_bzero(str, SIZEOF(str));\
+		snprintf(str, SIZEOF(str), format, __VA_ARGS__);\
+		ASSERT(strlen(str) + 1 < SIZEOF(str));\
+	}
+
+static inline bool areMoreScriptsExpected()
+{
+	// if the number of remaining scripts is not bigger than 0, then this request
+	// is invalid in the current context, as Ledger was not expecting another
+	// script to be parsed
+	return ctx->complexScripts[ctx->level].remainingScripts > 0;
+}
+
+static inline bool isComplexScriptFinished()
+{
+	return ctx->level > 0 && ctx->complexScripts[ctx->level].remainingScripts == 0;
+}
+
+static inline void complexScriptFinished()
+{
+	while (isComplexScriptFinished()) {
+		ASSERT(ctx->level > 0);
+		ctx->level--;
+
+		ASSERT(ctx->complexScripts[ctx->level].remainingScripts > 0);
+		ctx->complexScripts[ctx->level].remainingScripts--;
+
+		TRACE_WITH_CTX("complex script finished, ");
+	}
+}
+
+static inline void simpleScriptFinished()
+{
+	ASSERT(ctx->complexScripts[ctx->level].remainingScripts > 0);
+	ctx->complexScripts[ctx->level].remainingScripts--;
+
+	TRACE_WITH_CTX("simple script finished, ");
+
+	if (isComplexScriptFinished()) {
+		complexScriptFinished();
+	}
+}
+
+// UI
+
+const char ui_native_script_header[7][27] = {"Script - key path", "Script - key", "Script - ALL", "Script - ANY", "Script - N of K", "Script - invalid before", "Script - invalid hereafter"};
+
+#define ASSERT_UI_SCRIPT_TYPE_SANITY() ASSERT(ctx->ui_scriptType >= UI_SCRIPT_PUBKEY_PATH && ctx->ui_scriptType <= UI_SCRIPT_INVALID_HEREAFTER)
+#define HEADER ui_native_script_header[ctx->ui_scriptType]
+
+static uint8_t _getScriptLevelForPosition()
+{
+	// For complex scripts we reduce the current level by 1
+	// Because they already have the level increased by 1
+	int levelOffset = ctx->ui_scriptType == UI_SCRIPT_ALL
+	                  || ctx->ui_scriptType == UI_SCRIPT_ANY
+	                  || ctx->ui_scriptType == UI_SCRIPT_N_OF_K
+	                  ? -1 : 0;
+	ASSERT(ctx->level + levelOffset >= 0);
+	return ctx->level + levelOffset;
+}
+
+static void deriveScriptHash_display_ui_position(uint8_t level, ui_callback_fn_t* callback)
+{
+	ASSERT_UI_SCRIPT_TYPE_SANITY();
+	TRACE();
+
+	// For level n the longest possible position is: "x.x.x...x"
+	// where the x is 2^32-1 (length 10) and there are up to level xs
+	// + the leading "Position: "
+	char position[10 + 11 * (MAX_SCRIPT_DEPTH - 1) + 1];
+	int offset = snprintf(position, 10 + 1, "Position: ");
+	for (int i = 1; i <= level; i++) {
+		snprintf(&position[offset], 11 + 1, "%u.", ctx->complexScripts[i].totalScripts - ctx->complexScripts[i].remainingScripts + 1);
+		// snprintf returns 0, https://github.com/LedgerHQ/nanos-secure-sdk/issues/28
```

Comment:
In that case, then offset's initial value will also be 0, so "Position: " would be overwritten.

Reply:
For some reasons the `snprintf` function works like a black magic. With no formatting and snprinting a simple string, the return value of `snprintf` is correct. But when you include some formatting like `"%u"`, then it returns a 0.

Reply:
I can update the comment to reflect this.

Reply:
That would be helpful, because the linked issue doesn't make this obvious. Question remains if we can rely on this snprintf implementation for even not formatted text. Maybe an ASSERT to make sure offset starts at strlen("Position: ")?

Reply:
I'd ignore the return value of snprintf, it is not reliable, rather we should always use `strlen()` to find out what it did.

Also, the value `ctx->complexScripts[i].totalScripts - ctx->complexScripts[i].remainingScripts + 1` perhaps deserves to be stored in an aptly named variable.

And maybe it makes sense to somehow name the constants (what are 10, 11 etc.)?

Ah, and one more: before using `complexScripts[i]`, we should assert the index i is not out of bounds (the condition `i <= level` looks suspicious). Consider using size_t for the for loop variable as we do elsewhere (though maybe not with 100% consistency?).

And perhaps `position` is not a suitable name for a char array, what about something like `(script)PositionDescription`?

Reply:
I tried to address all your points.